crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -845,32 +842,56 @@ where
         self.valid_note_commitment_update(&shielded_tx)?;
 
         // Checks on the transparent bundle, if present
+        let mut changed_bals_minus_txn = changed_balances.clone();
         validate_transparent_bundle(
             &shielded_tx,
-            &mut changed_balances,
+            &mut changed_bals_minus_txn,
             masp_epoch,
             conversion_state,
-            &mut signers,
+            &mut authorizers,
         )?;
 
-        // Ensure that every account for which balance has gone down has
-        // authorized this transaction
-        for (addr, pre) in changed_balances.pre {
-            if changed_balances
-                .post
-                .get(&addr)
-                .unwrap_or(&ValueSum::zero())
-                < &pre
-                && addr != masp_address_hash
+        // Ensure that every account for which balance has gone down as a result
+        // of the Transaction has authorized this transaction
+        for (addr, minus_txn_pre) in changed_bals_minus_txn.pre {
+            // The pre-balance seen by all VPs including this one
+            let pre = changed_balances.pre.get(&addr).unwrap_or(&zero);
+            // The post-balance seen by all VPs including this one
+            let post = changed_balances.post.get(&addr).unwrap_or(&zero);
+            // The post-balance if the effects of the Transaction are removed
+            let minus_txn_post =
+                changed_bals_minus_txn.post.get(&addr).unwrap_or(&zero);
+            // Never require a signature from the MASP VP
+            if addr != masp_address_hash &&
+            // Only require further authorization if without the Transaction,
+            // this Tx would decrease the balance of this address
+                minus_txn_post < &minus_txn_pre &&
+            // Only require further authorization from this address if the
+            // Transaction alters its balance
+                (minus_txn_pre, minus_txn_post) != (pre.clone(), post)
             {
-                signers.insert(addr);
+                // This address will need to provide further authorization
+                authorizers.insert(addr);
             }
         }
 
+        let mut actions_authorizers: HashSet<&Address> = actions
```

Comment:
`Actions` are actually a vector of `Action`, so it's possible to push the same action more than once. I believe from the perspective of the masp vp this is of no importance: we just want to check that the action for a specific authorizer has been written, it doesn't matter if it's duplicated